File: apps/workbench/lib/bff/admin/integrations.ts
// Integrations status page read (0.0.4 S15, FR-23) for the Admin BFF (ADR-0093
// admin route group). Pure and dependency-injected like the sibling admin/*.ts
// modules; the thin app/api/admin/integrations route wraps this with withSession +
// createAdminApiClient.
//
// Dispatches over the SUPERVISOR ADMIN Profile API (deps.ts's createAdminApiClient),
// like admin/dead-letter.ts: toee_integrations is allowlisted for supervisor_admin
// (hermes/toee_hermes/plugin/profiles.py). The single action is agent-excluded on
// the Hermes side (_AGENT_EXCLUDED_ACTIONS) -- never reachable from a live agent's
// tool loop.
//
// ROLE GATING: this page + route are ADMIN-ONLY (lib/auth/access.ts), deliberately
// NARROWER than the supervisor+admin dead-letter operations view -- integrations are
// a CREDENTIAL surface (gap-review P4). The Hermes profile that answers the dispatch
// (supervisor_admin) is orthogonal to the workbench role gate; the admin-only check
// is enforced by withSession before this handler runs.
//
// READ, fail-open (dispatch, not dispatchWrite): a view needs no actor attribution.
// The handler NEVER fabricates a "healthy" -- the backend reports config presence
// honestly, last_successful_call stays null (nothing records it), and last_probe is
// the latest scheduled-probe result (S16, FR-24) or null when none has run yet (the
// panel renders "unknown" / "never probed" / the probe badge).
import type { HermesApiClient } from "../../gateway/hermes-api-client";
import { HermesApiError } from "../../gateway/hermes-api-client";
import { hermesErrorToProblem } from "../../gateway/hermes-error";
import { json } from "../respond";

// The latest scheduled health-probe result for one integration (S16, FR-24). The
// three honest states the probe records: "ok" (reachable + authorized), "failed"
// (credential present but the read errored/was ambiguous -- `reason` explains), and
// "not_configured" (no credential, so the probe was SKIPPED). `reason` is a short,
// secret-free string, non-null only on "failed". `checkedAt` is when it ran.
export interface ProbeResult {
  status: string;
  reason: string | null;
  checkedAt: string;
}

// One integration's uniform status. `configured` is the honest credential-presence
// signal (green only when the credential is actually set, not merely that the code
// path exists). `pinnedVersion` is a version string (Composio toolkits only), never
// a secret. `lastSuccessfulCall` is null everywhere in S15 (nothing records it yet)
// and `lastProbe` is the newest probe result (S16) or null when none has run yet --
// the panel shows "unknown"/"never probed"/the probe badge rather than a fabrication.
export interface IntegrationStatus {
  key: string;
  label: string;
  kind: string;
  configured: boolean;
  status: string;
  pinnedVersion: string | null;
  lastSuccessfulCall: string | null;
  lastProbe: ProbeResult | null;
  detail: string;
}

export interface IntegrationsView {
  // The active external-vendor backend (mock | composio). Composio credentials can
  // be present while the live backend is still `mock`, so this is shown alongside
  // the rows to keep "configured" honest about whether calls actually route live.
  activeDriver: string;
  integrations: IntegrationStatus[];
}

function malformed(detail: string): never {
  throw new HermesApiError(
    "unexpected_error",
    `malformed integrations payload: ${detail}`,
  );
}

function requireString(value: unknown, field: string): string {
  if (typeof value !== "string") malformed(field);
  return value as string;
}

function requireNullableString(value: unknown, field: string): string | null {
  if (value === null || value === undefined) return null;
  if (typeof value !== "string") malformed(field);
  return value as string;
}

function requireBoolean(value: unknown, field: string): boolean {
  if (typeof value !== "boolean") malformed(field);
  return value as boolean;
}

function requireObject(value: unknown, field: string): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    malformed(field);
  }
  return value as Record<string, unknown>;
}

function requireArray(value: unknown, field: string): unknown[] {
  if (!Array.isArray(value)) malformed(field);
  return value as unknown[];
}

// Strict, like mapIntegration: a null last_probe is a legitimate "never probed",
// but a present-but-malformed one throws (502) rather than being silently dropped
// or painted as a fabricated state.
function mapProbe(raw: unknown): ProbeResult | null {
  if (raw === null || raw === undefined) return null;
  const r = requireObject(raw, "last_probe");
  return {
    status: requireString(r.status, "last_probe.status"),
    reason: requireNullableString(r.reason, "last_probe.reason"),
    checkedAt: requireString(r.checked_at, "last_probe.checked_at"),
  };
}

function mapIntegration(raw: unknown): IntegrationStatus {
  const r = requireObject(raw, "integrations[]");
  return {
    key: requireString(r.key, "key"),
    label: requireString(r.label, "label"),
    kind: requireString(r.kind, "kind"),
    // Strict on purpose: a missing `configured` must NOT default to true, or a
    // backend shape change would paint an owner-blocked integration green.
    configured: requireBoolean(r.configured, "configured"),
    status: requireString(r.status, "status"),
    pinnedVersion: requireNullableString(r.pinned_version, "pinned_version"),
    lastSuccessfulCall: requireNullableString(
      r.last_successful_call,
      "last_successful_call",
    ),
    lastProbe: mapProbe(r.last_probe),
    detail: requireString(r.detail, "detail"),
  };
}

export function mapIntegrationsView(raw: unknown): IntegrationsView {
  const r = requireObject(raw, "root");
  return {
    activeDriver: requireString(r.active_driver, "active_driver"),
    integrations: requireArray(r.integrations, "integrations").map(mapIntegration),
  };
}

export async function handleGetIntegrationsStatusViaApi(
  client: HermesApiClient,
): Promise<Response> {
  try {
    const data = await client.dispatch(
      "toee_integrations",
      "get_integrations_status",
      {},
    );
    return json(mapIntegrationsView(data));
  } catch (err) {
    return hermesErrorToProblem(err);
  }
}

// --- S17 in-app reconnect (FR-25) --------------------------------------------
// Two shapes: Composio OAuth (initiate_reconnect -> a provider re-auth URL) and,
// for BOTH shapes' completion, an on-demand re-probe (reprobe_now). Both are
// governed WRITES: dispatchWrite, fail-closed on a missing actor. The acting admin
// rides HermesApiClient.actorAccountId from the signed-in session (ADR-0148) -- it
// is never a request param, and the routes never read one.

export interface ReconnectLink {
  integrationKey: string;
  redirectUrl: string;
}

export interface ReprobeReceipt {
  integrationKey: string;
  status: string;
  reason: string | null;
}

function mapReconnectLink(raw: unknown): ReconnectLink {
  const r = requireObject(raw, "root");
  // Strict: a null/absent redirect_url is a fail-closed backend (owner-blocked or a
  // wrong SDK guess), NOT a usable link -- surface it as an error, never navigate.
  return {
    integrationKey: requireString(r.integration_key, "integration_key"),
    redirectUrl: requireString(r.redirect_url, "redirect_url"),
  };
}

function mapReprobeReceipt(raw: unknown): ReprobeReceipt {
  const r = requireObject(raw, "root");
  return {
    integrationKey: requireString(r.integration_key, "integration_key"),
    status: requireString(r.status, "status"),
    reason: requireNullableString(r.reason, "reason"),
  };
}

// Only the three Composio-managed connections have an OAuth reconnect. Guarded here
// AND server-side (the handler rejects a non-Composio key), so a crafted body cannot
// coax a link for a static-token integration.
const COMPOSIO_KEYS = new Set(["shopify", "qbo", "square"]);

export async function handleInitiateReconnectViaApi(
  client: HermesApiClient,
  integrationKey: unknown,
  callbackUrl: string,
): Promise<Response> {
  if (typeof integrationKey !== "string" || !COMPOSIO_KEYS.has(integrationKey)) {
    return hermesErrorToProblem(
      new HermesApiError("not_found", "unknown Composio integration"),
    );
  }
  try {
    const data = await client.dispatchWrite(
      "toee_integrations",
      "initiate_reconnect",
      { integration_key: integrationKey, callback_url: callbackUrl },
    );
    return json(mapReconnectLink(data));
  } catch (err) {
    return hermesErrorToProblem(err);
  }
}

export async function handleReprobeNowViaApi(
  client: HermesApiClient,
  integrationKey: unknown,
): Promise<Response> {
  if (typeof integrationKey !== "string" || integrationKey === "") {
    return hermesErrorToProblem(
      new HermesApiError("not_found", "integrationKey is required"),
    );
  }
  try {
    const data = await client.dispatchWrite("toee_integrations", "reprobe_now", {
      integration_key: integrationKey,
    });
    return json(mapReprobeReceipt(data));
  } catch (err) {
    return hermesErrorToProblem(err);
  }
}

File: hermes-runtime/hermes_runtime/datastore/handlers/integrations.py
"""Datastore handler for ``toee_integrations`` -- the ``/admin/integrations``
status-page read (0.0.4 S15, FR-23).

**Admin-only, a credential surface.** ADR-0093 gives ``/admin/*`` supervisor+admin,
but the workbench narrows THIS page to admin-only (``lib/auth/access.ts``) --
deliberately narrower than the supervisor+admin dead-letter OPERATIONS view, because
integrations touch credentials (gap-review P4). Both actions are agent-excluded
(``_AGENT_EXCLUDED_ACTIONS``); this is reached only from the admin BFF's
deterministic ``tools:dispatch`` over the ``supervisor_admin`` profile.

**Honest config presence, no invented health.** Each integration's ``configured``
comes from that driver's OWN existing signal -- ``easyroutes_configured``,
``gadget_configured``, ``composio_config_status``, ``simpletexting_configured``,
``openrouter_configured`` -- read against THIS process's environment, which is the
same env the tool-executing processes share (docker-compose ``env_file``). So green
means the credential is actually present and a call would run, not merely that the
code path exists. An owner-blocked integration (no key yet) shows ``not_configured``,
never a fabricated ``healthy`` (the track's spine).

**No secret ever leaves.** Only booleans, the Composio version pin STRING (a version,
not a credential), and env-var NAMES are returned -- never a token/key/account-id
value. The repo secret-scan gate stays green (NFR-6).

**Honest until S17.**
- ``last_successful_call`` is ``None`` everywhere: nothing in the drivers records one
  yet, so the panel shows "unknown" rather than a fabricated timestamp.
- ``last_probe`` (S16, FR-24): the LATEST ``integration_probe`` row per integration,
  ``{status, reason, checked_at}`` or ``None`` when no probe has run yet (the panel
  renders "never probed"). ``status`` is one of ``ok`` / ``failed`` /
  ``not_configured`` -- the three honest states the scheduled probe records; the page
  badges off it. A probe result is a status + reason string only, never a secret.

``conn`` is the deliberate S15 seam S16 fills: :func:`_latest_probes` SELECTs the
newest probe row per integration here. S17's reconnect reads the same
per-integration ``key`` this shape already carries.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from psycopg.rows import dict_row
from toee_hermes.errors import ToolDriverError

from toee_hermes.drivers.base import resolve_integration_driver
from toee_hermes.drivers.composio.driver import (
    composio_config_status,
    initiate_composio_reconnect,
)
from toee_hermes.drivers.easyroutes.driver import (
    API_TOKEN_ENV as EASYROUTES_TOKEN_ENV,
    easyroutes_configured,
)
from toee_hermes.drivers.gadget import API_KEY_ENV as GADGET_KEY_ENV, gadget_configured

from hermes_runtime.openrouter import openrouter_configured
from hermes_runtime.simpletexting_reply import simpletexting_configured

from ._common import insert_audit, read_string

if TYPE_CHECKING:  # pragma: no cover - typing only
    from toee_hermes.tool_gate import ToolExecutionContext

# The three Composio-managed connections -- the only ones with an OAuth reconnect
# (FR-25). The static-token integrations (easyroutes/simpletexting/openrouter/gadget)
# have no OAuth: their "reconnect" is guided env rotation + re-probe, never a link.
_COMPOSIO_KEYS = frozenset({"shopify", "qbo", "square"})

_OPENROUTER_KEY_ENV = "OPENROUTER_API_KEY"
_SIMPLETEXTING_TOKEN_ENV = "SIMPLETEXTING_API_TOKEN"

# Human labels for the three Composio Layer-1 toolkits (keys match
# composio_config_status()).
_COMPOSIO_LABELS = {
    "shopify": "Shopify (Composio)",
    "qbo": "QuickBooks (Composio)",
    "square": "Square (Composio)",
}


def _entry(
    *,
    key: str,
    label: str,
    kind: str,
    configured: bool,
    detail: str,
    last_probe: dict[str, Any] | None = None,
    pinned_version: str | None = None,
) -> dict[str, Any]:
    """One integration row in the uniform status shape.

    ``last_successful_call`` is always ``None`` (nothing records it yet -> "unknown").
    ``last_probe`` is the newest scheduled-probe row for this integration (S16) or
    ``None`` when none has run yet ("never probed") -- never a fabrication.
    """
    return {
        "key": key,
        "label": label,
        "kind": kind,
        "configured": configured,
        "status": "configured" if configured else "not_configured",
        "pinned_version": pinned_version,
        "last_successful_call": None,
        "last_probe": last_probe,
        "detail": detail,
    }


def _missing(env_name: str) -> str:
    return f"Not configured: set {env_name} in the deployment env."


def _latest_probes(conn) -> dict[str, dict[str, Any]]:
    """Newest ``integration_probe`` row per integration key (S16 seam, FR-24).

    Returns ``{key: {status, reason, checked_at}}``. ``conn`` is ``None`` only in the
    pure config-presence unit tests (which pass no DB); there the page has no probe
    history to show, so every ``last_probe`` stays ``None`` ("never probed").
    """
    if conn is None:
        return {}
    with conn.cursor(row_factory=dict_row) as cur:
        cur.execute(
            """
            SELECT DISTINCT ON (integration_key)
                   integration_key, status, reason, checked_at
            FROM integration_probe
            ORDER BY integration_key, checked_at DESC
            """
        )
        rows = cur.fetchall()
    return {
        row["integration_key"]: {
            "status": row["status"],
            "reason": row["reason"],
            "checked_at": row["checked_at"].isoformat(),
        }
        for row in rows
    }


def _get_integrations_status(
    conn, params: dict[str, Any], context: "ToolExecutionContext"
) -> Any:
    # A read -> no actor required, no audit (parity with get_retention_status).
    del params, context

    active_driver = resolve_integration_driver()
    composio = composio_config_status()
    probes = _latest_probes(conn)

    integrations: list[dict[str, Any]] = []

    for key, label in _COMPOSIO_LABELS.items():
        status = composio[key]
        configured = bool(status["configured"])
        if configured:
            detail = f"Connected account present, pinned to {status['pinned_version']}."
        elif not status["api_key_present"]:
            detail = _missing("COMPOSIO_API_KEY")
        elif not status["connected"]:
            detail = _missing(str(status["account_env"]))
        else:
            detail = (
                f"Connected but no version pin: set {status['version_env']} to an "
                "exact toolkit version (not 'latest')."
            )
        integrations.append(
            _entry(
                key=key,
                label=label,
                kind="composio_toolkit",
                configured=configured,
                pinned_version=status["pinned_version"],  # type: ignore[arg-type]
                detail=detail,
                last_probe=probes.get(key),
            )
        )

    er_configured = easyroutes_configured()
    integrations.append(
        _entry(
            key="easyroutes",
            label="EasyRoutes",
            kind="easyroutes",
            configured=er_configured,
            detail=(
                "Delivery m2m credentials present."
                if er_configured
                else _missing(f"{EASYROUTES_TOKEN_ENV} and EASYROUTES_CLIENT_ID")
            ),
            last_probe=probes.get("easyroutes"),
        )
    )

    st_configured = simpletexting_configured()
    integrations.append(
        _entry(
            key="simpletexting",
            label="SimpleTexting",
            kind="simpletexting",
            configured=st_configured,
            detail=(
                "Outbound SMS token present."
                if st_configured
                else _missing(_SIMPLETEXTING_TOKEN_ENV)
            ),
            last_probe=probes.get("simpletexting"),
        )
    )

    or_configured = openrouter_configured()
    integrations.append(
        _entry(
            key="openrouter",
            label="OpenRouter",
            kind="openrouter",
            configured=or_configured,
            detail=(
                "Model API key present."
                if or_configured
                else _missing(_OPENROUTER_KEY_ENV)
            ),
            last_probe=probes.get("openrouter"),
        )
    )

    gadget_ok = gadget_configured()
    integrations.append(
        _entry(
            key="gadget",
            label="Gadget mapping endpoint (paymentstatussync)",
            kind="gadget",
            configured=gadget_ok,
            detail=(
                "QBO<->Shopify mapping API key present."
                if gadget_ok
                else _missing(GADGET_KEY_ENV)
            ),
            last_probe=probes.get("gadget"),
        )
    )

    return {"active_driver": active_driver, "integrations": integrations}


def _require_actor(context: "ToolExecutionContext") -> str:
    """The acting admin for a governed reconnect action, or a denial (ADR-0148).

    The actor rides ``ToolExecutionContext.user_id``, asserted by the admin BFF from
    the signed-in session under the shared bearer -- NEVER a request param. Fail
    closed: an unattributed reconnect would act on a credential surface and write a
    NULL-actor audit row while reporting success. (Mirrors ``dead_letter._require_actor``.)
    """
    actor = context.user_id
    if not actor:
        raise ToolDriverError(
            "policy_blocked", "A governed reconnect requires an attributed actor."
        )
    return actor


def _initiate_reconnect(
    conn, params: dict[str, Any], context: "ToolExecutionContext"
) -> Any:
    """Composio OAuth reconnect (FR-25): generate the re-auth link, attributed + audited.

    Order: actor first (fail closed), then key validation, then the live SDK call --
    so an unattributed or invalid request leaves the audit log untouched. The link
    generation and the audit row share the handler's single transaction (the driver
    commits around the whole call), so a reconnect is never audited without the link
    actually being issued, and a failed SDK call (owner-blocked / wrong guess) rolls
    back with no audit row and a governed error.

    The ``callback_url`` is built server-side by the workbench route from its OWN
    origin + a state bound to the admin session; it is NOT client-controlled. No token
    is handled here -- Composio holds the credentials; this returns only a redirect URL.
    """
    actor = _require_actor(context)
    key = read_string(params, "integration_key", "integrationKey")
    if not key:
        raise ToolDriverError("unexpected_error", "integration_key is required.")
    if key not in _COMPOSIO_KEYS:
        raise ToolDriverError(
            "unexpected_error",
            f"'{key}' has no OAuth reconnect (static-token integrations rotate via env + re-probe).",
        )
    callback_url = read_string(params, "callback_url", "callbackUrl")
    if not callback_url:
        raise ToolDriverError("unexpected_error", "callback_url is required.")

    redirect_url = initiate_composio_reconnect(key, callback_url=callback_url)

    insert_audit(
        conn,
        profile=context.profile,
        account_id=actor,
        action="integration_reconnect_initiated",
        target_type="integration",
        target_id=key,
        details={"toolkit": key},
    )
    return {"integration_key": key, "redirect_url": redirect_url}


def _reprobe_now(conn, params: dict[str, Any], context: "ToolExecutionContext") -> Any:
    """On-demand health re-probe of ONE integration (FR-25), attributed + audited.

    Both reconnect shapes end here: the Composio OAuth return and the static-token env
    rotation both want a fresh badge NOW rather than on the next scheduled cycle. Runs
    the SAME per-integration probe wiring the scheduled job uses (same configured gate,
    same authenticated read, same three honest states) and records the row on the
    handler's own cursor -- so the probe row and the audit row commit together in the
    driver's single unit of work. A re-probe cannot report anything the scheduled probe
    wouldn't: an owner-blocked integration still records ``not_configured``, a live
    fault still records ``failed``, never a fabricated ``ok``.
    """
    from hermes_runtime.integration_probe import probe_one, write_probe_rows

    actor = _require_actor(context)
    key = read_string(params, "integration_key", "integrationKey")
    if not key:
        raise ToolDriverError("unexpected_error", "integration_key is required.")
    try:
        result = probe_one(key)
    except KeyError as err:
        raise ToolDriverError("unexpected_error", f"unknown integration '{key}'.") from err

    with conn.cursor() as cur:
        write_probe_rows(cur, [result])

    insert_audit(
        conn,
        profile=context.profile,
        account_id=actor,
        action="integration_reprobed",
        target_type="integration",
        target_id=key,
        details={"status": result.status},
    )
    return {"integration_key": key, "status": result.status, "reason": result.reason}


def integrations_handlers() -> dict[str, dict[str, Any]]:
    """Registry fragment for the integrations status-page read + S17 reconnect actions."""
    return {
        "toee_integrations": {
            "get_integrations_status": _get_integrations_status,
            "initiate_reconnect": _initiate_reconnect,
            "reprobe_now": _reprobe_now,
        }
    }


__all__ = ["integrations_handlers"]
